### PR TITLE
Annotate `Players` in ViewReport

### PR DESCRIPTION
**Pull request by GreenAsJade**
Fixes Moderators losing track of who is who

## Proposed Changes

  - Have  `Player` provide a context that it can be rendered in to enable it to annotate participants in a report.
  - Have `ViewReport` use that context.

### CODE REVIEW

**Comment by GreenAsJade** [comment]
Screenshot 2024-08-24 at 9 34 30 PM (2)

File: src/components/Player/Player.tsx [hunk 1]
```diff
@@ -283,6 +292,14 @@ export function Player(props: PlayerProperties): JSX.Element {
         main_attrs.className += " " + combined.ui_class;
     }
 
+    if (viewReportContext && viewReportContext.reported.id === player_id) {
```
Comment (anoek) [inline src/components/Player/Player.tsx:295]: This feels wrong, we shouldn't be spreading the context logic over to a generic component like `Player` I don't think. How about instead we put that context stuff up in the report code and then pass `"reported"` in for the `ui_class` for `Player`?

**Comment by GreenAsJade** [comment]
Here's my case for doing it this way :)

"Player" is a component that knows how to render Players in various contexts.

One of those is "in a report context".  It makes a context provider available to components that feel they are "in a report context", and lets those components describe which two players have the context-specific roles.

Once that's done neither component has to worry anymore about remembering to (or even being able to) think about what is the current' player's role.  Instead _every instance_ of `Player` within that context is "magically" annotated.

Note how the `ReportedGame` is made of a hiearachy of components, many of which have `Player` components instantiated in them, and every single one is correctly annotated, in the screenshot above, without editing any code in any of them...

**Comment by anoek** [comment]
> "Player" is a component that knows how to render Players in various contexts.

Does it? I always viewed it as a generic component that didn't care where it was rendered, it just rendered the bits of a player name that you told it to.

That changes fundamentally with this. Now part of the rendering code for a view is pushed down into a generic component, feels like that is a pattern that could get out of hand very quickly.

However, I could see how getting the data to all the right places might get painful. Let me ponder, will probably merge as is but want to digest if there's a cleaner way.

**Comment by anoek** [comment]
> >    Now part of the rendering code for a view is pushed down into a generic component

> I'm not seeing this. A Player component is responsible for rendering the Player - it only has Player rendering code.

ehh, the `Player.styl` now has code that only takes effect when in the `#ViewReport` element and we have a special `ViewReportContextType` type that is, as indicated by the variable name, clearly only applicable for the `ViewReport` view.

I guess though it probably wouldn't have felt so wrong if it was spun as a more generic thing, as opposed to something that only works when in `ViewReport`, so in that light maybe it's fine, and maybe if we find another place where that pattern is a good fit we make it more generic or something.

**Comment by GreenAsJade** [comment]
I think that there should be a contract about the styling.  The component that does instantiates the context

`  <ViewReportContext.Provider...`

should also keep the contract by putting a className that `Player` is expecting around the context as well, rather than `Player` looking for `#ViewReport`  (which is a specific special component).

I'll see about making it feel more like a generic pattern and then we'll see how it strikes you...

... and in the mean time if you can think of a different way of ensuring that "all `Player`s within this page get annotated" that'd be cool!

**Comment by GreenAsJade** [comment]
It felt safer to have a deal with the context consumer to specifically contain effects of the annotation styling.

**Pull request merged**